### ISSUE TITLE
Possibility of generating OpenAPI spec for `http:Request`, `http:Response` and `default` accessor

### ISSUE THREAD

**Issue by shafreenAnfar** [ballerina-platform/openapi-tools#836]
At the moment OpenAPI tool skip generating OpenAPI spec for resources which has at least one of the above. It does this by giving a warning.  The reason for this is the tool has no way of figuring out what the user meant by those. In other words, in order to generate the spec, the API must be typed as follows.

```ballerina
resource function put reservation(@http:Payload rep:Reservation reservation) 
                returns rep:ReservationCreated|rep:ReservationConflict|rep:SnowpeakInternalError
```

However, in cases such as the following, the tool skips the generation and prints a warning. 
```ballerina
resource function 'default passthrough(http:Request req) returns http:Response|error?
```
This issue is to discuss if there is a mapping for those constructs and OpenAPI constructs.

**Comment by shafreenAnfar** [comment]
For `http:Request` and `http:Response`, I think we can do a decent mapping using OpenAPI constructs as follows.

```yaml
requestBody:
  description: Any request
  content:
    "*/*":
      example:
        summary: "Any type of entity body"
responses:
  "default":
    description: Any response
    content:
      "*/*":
        example:
          summary: "Any type of entity body"
``` 

@hevayo @lnash94 any thoughts ?

**Comment by shafreenAnfar** [comment]
Consider the below example,

```ballerina
import ballerina/http;

service / on new http:Listener(9090) {
    resource function get hello(http:Request request) returns @http:Payload http:Response {
        return new http:Response();
    }
}
```
As per the above suggestion the complete OpenAPI spec would look like below.
```yaml
openapi: 3.0.1
info:
  title: Headers Openapi Yaml
  version: 0.0.0
servers:
- url: "{server}:{port}/"
  variables:
    server:
      default: http://localhost
    port:
      default: "9090"
paths:
  /hello:
    get:
      operationId: operation_get_/hello
      requestBody:
        description: Any request
        content:
          "*/*":
            example:
              summary: "Any type of entity body"
      responses:
        "default":
          description: Any Reponse
          content:
            "*/*":
              example:
                summary: "Any type of entity body"
components: {}
```

**Comment by indikasampath2000** [comment]
I could see an issue in the above OAS definition. GET operation cannot have a requestBody. It is giving parsing error. 

I have modified it to a POST and generated the client. We have mapped `content: "*/*"` to `byte[]` for the request because it is considered as Multiple Media Types. The return is `http:Response`.

```
import ballerina/http;

public isolated client class Client {
    final http:Client clientEp;
    # Gets invoked to initialize the `connector`.
    #
    # + clientConfig - The configurations to be used when initializing the `connector` 
    # + serviceUrl - URL of the target service 
    # + return - An error if connector initialization failed 
    public isolated function init(http:ClientConfiguration clientConfig =  {}, string serviceUrl = "http://localhost:9090/") returns error? {
        http:Client httpEp = check new (serviceUrl, clientConfig);
        self.clientEp = httpEp;
        return;
    }
    #
    # + payload - Any request 
    # + return - Any Reponse 
    remote isolated function operationGetHello(byte[] payload) returns http:Response|error {
        string resourcePath = string `/hello`;
        http:Request request = new;
        request.setPayload(payload, "*/*");
        http:Response response = check self.clientEp->post(resourcePath, request);
        return response;
    }
}
```

**Comment by shafreenAnfar** [comment]
> I could see an issue in the above OAS definition. GET operation cannot have a requestBody. It is giving parsing error.

Yeah, you are right, that was a mistake.

**Comment by shafreenAnfar** [comment]
Interesting, I look at it differently. 

`*/*` is basically means the media type could be anything. It could be `application/json`, `application/xml`, `text/plain` and the list goes on. Now, to keep things simple if you just consider the request side of the service, this basically means service accepts any type of request. Therefore, given this is open-ended, it is better to map `*\*` to `http:Request`. `http:Request` allows users to deal with any type of request.

Therefore, when you generate the client side, I believe the signature should be as follows.

```ballerina
remote isolated function operationGetHello(http:Request request) returns http:Response|error
```
Since the service could receive any type of request, the client should be able to send any type of request. As I see it having a `byte[]` instead is problematic. Because usually `byte[]` simply translates into `application/octet-stream`.  Therefore, the only type of request the user could send is `application/octet-stream`. Also, it seems the generated code is a bit wrong.

```ballerina
request.setPayload(payload, "*/*");
```
Because there is no such mime type (`*/*`). In this case since the `payload` is `byte[]`, the mime type should be `application/octet-stream`.

**Comment by lnash94** [comment]
> ```yaml
> "default":
>           description: Any Reponse
>           content:
>             "*/*":
>               example:
>                 summary: "Any type of entity body"
> ```

Yeah, we can give the wild card "*/*" for the response and the request body that we can't map without many details. I checked the OAS provide and it is valid for having a wildcard. 😃

**Comment by aneeshafedo** [comment]
> ```ballerina
> request.setPayload(payload, "*/*");
> ```
>   
> Because there is no such mime type (*/*). In this case since the payload is byte[], the mime type should be application/octet-stream.  And depending the payload the user should be able to set the right mime type (some of them are set by default).
> 

Yes, this is true in the current openapi to client generation we generate this line as below. 

```ballerina
request.setPayload(payload, "application/octet-stream");
```

**Sample OpenAPI**

```yaml
requestBody:
   content:
         */*:
           schema:
             type: string
             format: binary
```

**Generated remote function**
```ballerina
remote isolated function createPet(byte[] payload) returns http:Response|error {
       string path = string `/pets`;
       http:Request request = new;
       request.setPayload(payload, "application/octet-stream");
       http:Response response = check self.clientEp->post(path, request, targetType = http:Response);
       return response;
   }

```

This was generated referencing the following instruction given in the OpenAPI spec. 
```
 */* represents all types and is functionally equivalent to application/octet-stream
```

Refer [here](https://swagger.io/docs/specification/describing-request-body/#:~:text=*/*%20represents%20all%20types%20and%20is%20functionally%20equivalent%20to%20application/octet%2Dstream) for more details.

**Comment by lnash94** [comment]
- [ ] `http:Request` improvement
- [ ] `http:Response` improvement
- [ ] `defualt` method improvement

**Comment by shafreenAnfar** [comment]
@aneeshafedo yes, in the case of `application/octet-stream` you don't know what is in it the byte stream it could be anything. It is the same case as using `*/*`. Because from the service point of view, it doesn't know what could come in. 

But because they are functionally equivalent we don't have to restrict the user for only using `application/octet-stream`. That significantly reduces the user experience. Because we force the user to send `application/octet-stream` when the service don't really care about what is coming in. Functionally equivalent doesn't really mean both things are identical.

In addition, in Ballerina there is a clear mapping for `application/octet-stream` which is `byte[]`. Therefore, if the user wants to write a service that only accepts a `byte[]`, he/she could use the `byte[]` in the place of `http:Request`.  Therefore, the better mapping would be for the `http:Request` is `*/*`.

Lastly, the API design itself is not consistent. `*/*` when used for the response it generates `http:Response` but `*/*` when used for the request it generates a `byte[]`.

**Comment by shafreenAnfar** [comment]
The OpenAPI sample you have provided is not the same as the sample originally proposed. Those are two different things. 

Your sample
```yaml
requestBody:
   content:
       */*:
         schema:
           type: string
           format: binary
```
Originally proposed sample
```yaml
requestBody:
  description: Any request
  content:
    "*/*":
      example:
        summary: "Any type of entity body"
```
In the case of your sample, generating `byte[]` for input is correct. Because it says it accepts any type of byte stream but not anything else. Which is in the HTTP world maps to `application/octet-stream`. Therefore, in the *context* of your sample `*/*` is functionally equivalent to `application/octet-stream`. 

But those are two different things.

**Comment by shafreenAnfar** [comment]
Yeah, that is correct. However, I think we are not there yet to handle everything with `anydata`. For instance multipart requests/responses or large file upload request/response.

**Comment by shafreenAnfar** [comment]
Reopened this issue as we are yet to come to a conclusion on `default` accessor.
